Handle quantifiers with statement expressions

qinheping opened the pull request:
This pull request addresses the issue of handling quantifiers that contain statement expressions in CBMC. This fix ensures that CBMC can correctly process and verify programs that use quantifiers with statement expressions.

### Changes

- `goto_clean_expr.cpp`: Added logic to handle quantifiers with statement expressions. This includes:
  - A new function `find_base_symbol` to identify the base symbol in an expression.
  - Modifications to `clean_expr` to handle quantifiers with statement expressions by converting them pure expressions.
- `c_typecheck_expr.cpp`: Updated to allow quantifiers with statement expressions.
- Tests: Introduced new tests to verify the correct handling of quantifiers with statement expressions. These tests ensure that CBMC can process and verify programs that use such quantifiers without errors.

### Motivation and Context
Statement expressions may not contain side effects, and should be accepted as quantifier bodies when they don't.
Supporting quantifiers with statement expressions enhances CBMC's capability to handle a broader range of programs, e.g., GOTO program compiled from Rust MIR https://github.com/model-checking/kani/pull/3737.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening commented:
I am sure it makes sense to extend what's allowed in quantifiers. I am surprised by the choice of a (fragment) of statement expressions, which are a gcc extension at the C language level. In particular, note that this means you can't actually use these in Kani.

The examples suggest that you are perhaps looking for let expressions? These are unproblematic to use in quantifiers, and Kani can use these right away.

qinheping commented:
> In particular, note that this means you can't actually use these in Kani.

Kani is actually the motivation of support statement expressions in quantifiers. In Kani, we generate GOTO programs from Rust MIR. However, all Rust expressions will be compiled into a sequence of statements (including decl, assign, and conditional goto) in Rust MIR. We typically wrap all such statements into a closure function (in Rust) so that the evaluation of the Rust expressions becomes a single function call in GOTO. This approach doesn't work here because function calls are also side-effect expressions in GOTO. Therefore, we are considering inlining all function calls in MIR and generating a statement expression in GOTO from the inlined expression.

I don't think let-expressions are enough here because there will still be function calls or statement expressions in the where clauses of the let-expressions.

tautschnig commented on regression/cbmc/Quantifiers1/quantifier-with-function-call.desc:
```diff
@@ -0,0 +1,9 @@
+CORE
+quantifier-with-function-call.c
+
+^EXIT=134$
```
You'll have to add this one to `regression/validate-trace-xml-schema/check.py` as the failing invariant check implies the overall output isn't valid XML. But if you pick up my suggestion below then this will turn into `EXIT=6` and valid XML will be produced.

tautschnig commented on regression/cbmc/Quantifiers1/quantifier-with-side-effect.desc:
```diff
@@ -1,8 +1,8 @@
 
-^EXIT=6$
+^EXIT=134$
 ^SIGNAL=0$
-^file .* line 10 function main: quantifier must not contain side effects$
+^Invariant check failed
+^Reason: quantifier must not contain side effects
```
You'll have to add this one to `regression/validate-trace-xml-schema/check.py` as the failing invariant check implies the overall output isn't valid XML. But if you pick up my suggestion below then we're back with `EXIT=6` and you just need to update the error message ("quantifier must not contain function call").

kroening commented:
> Kani is actually the motivation of support statement expressions in quantifiers. In Kani, we generate GOTO programs from Rust MIR. However, all Rust expressions will be compiled into a sequence of statements (including decl, assign, and conditional goto) in Rust MIR. We typically wrap all such statements into a closure function (in Rust) so that the evaluation of the Rust expressions becomes a single function call in GOTO. This approach doesn't work here because function calls are also side-effect expressions in GOTO. Therefore, we are considering inlining all function calls in MIR and generating a statement expression in GOTO from the inlined expression.
> 
> I don't think let-expressions are enough here because there will still be function calls or statement expressions in the where clauses of the let-expressions.

Ok, then how about defining the concept of a "pure" function, and then allowing those in quantifiers?

We need that in the front-end anyway, for C++'s `constexpr`.

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -316,10 +316,10 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
 
-    if(has_subexpr(where, ID_side_effect))
+    if(expr.id() == ID_lambda && has_subexpr(where, ID_side_effect))
     {
       error().source_location = expr.source_location();
-      error() << "quantifier must not contain side effects" << eom;
+      error() << "lambda expressions must not contain side effects" << eom;
       throw 0;
     }
```
I suggest to do a bit more work here: let's also catch function calls, which should be doable by using
````c++
if(has_subexpr(where, [&](const exprt &subexpr) {
       return can_cast_expr<side_effect_exprt>(subexpr) &&
              can_cast_expr<side_effect_expr_function_callt>(subexpr);
    }))
{
  error().source_location = expr.source_location();
  error() << "quantifier must not contain function calls" << eom;
  throw 0;
}
````

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+           ID_statement_expression),
+      "quantifier must not contain side effects");
+
+    // Handle the case that quantifier body is a statement expression.
+    if(
+      code.operands()[0].id() == ID_side_effect &&
+      code.operands()[0].get_named_sub()[ID_statement].id() ==
+        ID_statement_expression)
+    {
```
Can you please move the code that's in this if branch to a method of its own? It has a very specific purpose and the overall code will be much easier to read when the `clean_expr` code can just invoke a named method.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
     DATA_INVARIANT(
-      !has_subexpr(expr, ID_side_effect),
-      "the front-end should check quantified expressions for side-effects");
+      !has_subexpr(expr, ID_side_effect) ||
+        (code.operands()[0].id() == ID_side_effect &&
+         code.operands()[0].get_named_sub()[ID_statement].id() ==
+           ID_statement_expression),
+      "quantifier must not contain side effects");
```
Hmm, this one permits arbitrary side effects as long as they are encapsulated in a statement expression. IMHO the only permissible side effect should be statement expressions, so this check should apply recursively. Please use a lambda as suggested for the function-call test above.

qinheping replied:
My idea in this PR is to convert all the side-effect expressions (except for function calls) in the quantifier body first, and then check that if the expression is side-effect free by checking if all the lhs of assigns are local declared. Let me double check if there could be other side-effect expression in MIR.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+        natural_loops.loop_map.size() == 0,
+        "quantifier must not contain side effects");
+
+      // `last` is the instruction corresponding to the last expression in the
+      // statement expression.
```
Note to self: continue the review from here.

tautschnig reviewed (commented):
A few initial comments, still need to work through the core piece of the implementation (see my "Note to self").

tautschnig commented:
> We need that in the front-end anyway, for C++'s `constexpr`.

But then `constexpr` has C++-standards-version-specific peculiarities, so a single definition might not work? I had pointed @qinheping to my `constexpr` implemention in https://github.com/diffblue/cbmc/pull/8386 as an initial guidance, though.

qinheping commented:
> Ok, then how about defining the concept of a "pure" function, and then allowing those in quantifiers?

In the design of this PR, no function calls are allowed in quantifiers.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
   else if(expr.id() == ID_forall || expr.id() == ID_exists)
   {
+    quantifier_exprt &qex = to_quantifier_expr(expr);
+    code_expressiont code{qex.where()};
```
Are you intentionally taking a copy here? I would like to think a const reference should do?

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+      natural_loops_mutablet natural_loops(where);
+      INVARIANT(
+        natural_loops.loop_map.size() == 0,
+        "quantifier must not contain side effects");
```
That's not really the right error message, is it? It should be about not permitting loops.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+
+      // `last` is the instruction corresponding to the last expression in the
+      // statement expression.
+      goto_programt::const_targett last;
```
I'd suggest that you initialise this to `where.instructions.end()` and, after the loop, have a `DATA_INVARIANT` that `last != where.instructions.end()`. As is, `last->get_other()` could potentially be using an uninitialised iterator.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+      }
+      auto last_expr =
+        expr_try_dynamic_cast<code_expressiont>(last->get_other())
+          ->expression();
```
I think this is a case where you're better off with `to_code_expressiont` instead of `expr_try_dynamic_cast` as you don't have a way to handle when that conversion fails (i.e., you really want to fail an invariant at that point).

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+        expr_try_dynamic_cast<code_expressiont>(last->get_other())
+          ->expression();
+
+      // `pathes` contains all the `targett` we are iterating over.
```
I think the modern form is "paths", apparently "pathes" is Middle English.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+
+      // `pathes` contains all the `targett` we are iterating over.
+      std::vector<goto_programt::const_targett> pathes;
+      pathes.push_back(where.instructions.begin());
```
1. How can a single vector of instruction iterators represent multiple execution paths?
2. In the spirit of RAII: `std::vector<goto_programt::const_targett> paths{1, where.instructions.begin()};`

qinheping replied:
1. The vector store the next target we are going to visit in each path.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+        std::make_pair(true_exprt(), replace_symbolt()));
+
+      std::unordered_set<symbol_exprt, irep_hash> declared_symbols;
+      // All bounded variables are local.
```
These are "bound" variables, not "bounded" ones.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+
+      exprt res = true_exprt();
+
+      // Visit the quantifeir body along `pathes`.
```
```suggestion
      // Visit the quantifier body along `pathes`.
```

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+      // `pathes` contains all the `targett` we are iterating over.
+      std::vector<goto_programt::const_targett> pathes;
+      pathes.push_back(where.instructions.begin());
+      std::vector<std::pair<exprt, replace_symbolt>>
```
Is it the case that this has to have exactly the same number of entries as `pathes`? Perhaps you want to introduce a local `struct` that has three members and then just have a single `std::vector` that holds elements of that struct type?

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+          "Quantifier body must have a unique end expression.");
+
+        // Add all local-declared symbols into `declared_symbols`.
+        if(current_it->is_decl())
```
Should you perhaps use a `switch` statement here instead?

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
+          exprt rhs = current_it->assign_rhs();
+          INVARIANT(
+            !has_subexpr(rhs, ID_function_call),
+            "quantifier must not contain side effects");
```
As `rhs` has been subject to all rules of goto conversion (and its cleaning) there really shouldn't ben any kind of side effect left on the right-hand side of an assignment. I believe this `INVARIANT` is quite possibly pointless as `ID_function_call` expressions can't possibly exist at that point.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+    std::make_pair(true_exprt(), replace_symbolt()));
+
+  std::unordered_set<symbol_exprt, irep_hash> declared_symbols;
+  // All bounded variables are local.
```
```suggestion
  // All bound variables are local.
```

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+
+  exprt res = true_exprt();
+
+  // Visit the quantifeir body along `pathes`.
```
```suggestion
  // Visit the quantifier body along `pathes`.
```

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+  auto last_expr =
+    expr_try_dynamic_cast<code_expressiont>(last->get_other())->expression();
+
+  // `pathes` contains all the `targett` we are iterating over.
```
I think modern English is "paths", "pathes" is apparently Middle English.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+    expr_try_dynamic_cast<code_expressiont>(last->get_other())->expression();
+
+  // `pathes` contains all the `targett` we are iterating over.
+  std::vector<goto_programt::const_targett> pathes;
+  pathes.push_back(where.instructions.begin());
+  std::vector<std::pair<exprt, replace_symbolt>> path_conditions_and_value_maps;
+  path_conditions_and_value_maps.push_back(
+    std::make_pair(true_exprt(), replace_symbolt()));
```
1. Isn't it the case that the two vectors must have a matching number of entries? I suggest you introduce a `struct` with three members and have a single vector holding elements of that struct.
2. RAII: `std::vector<goto_programt::const_targett> pathes{1, where.instructions.begin()};`

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+      "Quantifier body must have a unique end expression.");
+
+    // Add all local-declared symbols into `declared_symbols`.
+    if(current_it->is_decl())
```
You might want to use a `switch` statement here.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+        !has_subexpr(rhs, ID_function_call),
+        "quantifier must not contain side effects");
+      value_map.replace(rhs);
+      value_map.set(to_symbol_expr(current_it->assign_lhs()), rhs);
```
1. Your `find_base_symbol` function supports all kinds of expressions, but then here you unconditionally do `to_symbol_expr`?!
2. Use `lhs` instead of `current_it->assign_lhs()`.

qinheping replied:
I use `replace_expr` instead.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+        declared_symbols.count(find_base_symbol(lhs)),
+        "quantifier must not contain side effects");
+      exprt rhs = current_it->assign_rhs();
+      INVARIANT(
+        !has_subexpr(rhs, ID_function_call),
+        "quantifier must not contain side effects");
```
I believe this `INVARIANT` isn't as useful: I don't think `ID_function_call` could ever appear at this point (it would be a side effect with a statement of id "function_call"), and goto conversion with its cleaning really must have done away with any kind of side effect at this point already.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+    // path_condition && cond.
+    if(current_it->is_goto())
+    {
+      if(current_it->has_condition())
```
I don't think this does what you want here: `has_condition()` will always be `true` when `is_goto()` is `true`. Perhaps you wanted `!current_it->condition().is_true()`?

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+      continue;
+    }
+    // EXPRESSION(expr)
+    // The last instruction is an exression statement.
```
```suggestion
    // The last instruction is an expression statement.
```

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+      pathes.pop_back();
+      continue;
+    }
+    // DEAD symbol
+    // Remove symbol from declared_symbols.
+    if(current_it->is_dead())
+    {
+      declared_symbols.erase(current_it->decl_symbol());
+    }
```
I don't think you need to bother: symbol names are unique, so even an assignment to a variable that has been marked dead still refers to that local variable and cannot possibly be about a side effect.

tautschnig commented on src/ansi-c/goto-conversion/goto_clean_expr.cpp:
```diff
@@ -16,9 +16,172 @@ Author: Daniel Kroening, kroening@kroening.com
+    {
+      declared_symbols.erase(current_it->decl_symbol());
+    }
+    // Unsupported instructions.
+    if(current_it->is_function_call())
+    {
+      UNREACHABLE;
+    }
```
The switch statement would really help you here as it would make sure you intentionally handle/not handle all possible kinds of statements.

qinheping committed:
Fix: Handle quantifiers with statement expressions in CBMC

This commit addresses the issue of handling quantifiers that contain statement expressions in CBMC. Previously, quantifiers without side effects were not supported.

Changes include:
- Added logic to handle quantifiers with statement expressions in `goto_clean_expr.cpp`.
- Updated `c_typecheck_expr.cpp` to allow quantifiers with statement expressions.
- Introduced new tests to verify the correct handling of quantifiers with statement expressions.
